essence-of-live-coding-vivid/src/LiveCoding/Vivid.hs

```diff
@@ -0,0 +1,100 @@
+  | Stopped
+  deriving (Eq)
+
+{- | A 'ParametrisedHandle' corresponding to one @vivid@/@SuperCollider@ synthesizer.
+
+Usually, you will want to use 'liveSynth' instead, it is easier to handle.
+-}
+vividHandleParametrised
+  :: (VividAction m, Eq params, VarList params, Subset (InnerVars params) args, Elem "gate" args)
+  => ParametrisedHandle (params, SynthDef args, SynthState) m (Maybe (Synth args))
+vividHandleParametrised = ParametrisedHandle { .. }
+  where
+    createParametrised (params, synthDef, Started) = Just <$> synth synthDef params
+    createParametrised (params, synthDef, Stopped) = defineSD synthDef >> pure Nothing
+
+    destroyParametrised _ synthMaybe = traverse_ release synthMaybe
+
+    -- Only the synth parameters changed and it's still running.
+    -- So simply set new parameters without stopping it.
+    changeParametrised (paramsOld, synthDefOld, Started) (paramsNew, synthDefNew, Started) (Just synth)
+      | paramsOld /= paramsNew && synthDefOld == synthDefNew = do
+          set synth paramsNew
+          return $ Just synth
+    -- Synthdef or start/stop state changed, need to release and reinitialise
+    changeParametrised old new synth = defaultChange createParametrised destroyParametrised old new synth
+
+deriving instance Eq (SynthDef args)
+deriving instance Data SynthState
+deriving instance KnownSymbol a => Data (I a)
+
+{- | Create a synthesizer.
+
+When you add 'liveSynth' to your live program,
+it will be started upon reload immediately.
+
+Feed the definition of the synthesizer and its current intended state to this cell.
+The input has the form @(params, sdbody, synthState)@.
+
+* A change in @params@ will reload the synthesizer quickly,
+  unless the types of the parameters change.
+* A change in the 'sdbody :: SDBody' ...' or the _types_ of the @params@ will 'release' the synthesizer and start a new one.
+* The input 'synthState :: SynthState' represent whether the synthesizer should currently be running or not.
+  Changes in it quickly start or stop it.
+  * When it is started, @'Just' synth@ is returned, where @synth@ represents the running synthesizer.
+  * When it is stopped, 'Nothing' is returned.
+
+For an example, have a look at the source code of 'sine'.
+-}
+liveSynth
+  :: ( VividAction m
+     , Eq params, Typeable params, VarList params
+     , Typeable (InnerVars params), Subset (InnerVars params) (InnerVars params)
+     , Elem "gate" (InnerVars params), Data params
+     )
+  => Cell (HandlingStateT m)
+       (params, SDBody' (InnerVars params) [Signal], SynthState)
+       (Maybe (Synth (InnerVars params)))
+liveSynth = proc (params, sdbody, synthstate) -> do
+  paramsFirstValue <- hold -< params
+  handlingParametrised vividHandleParametrised -< (params, sd paramsFirstValue sdbody, synthstate)
+
+-- | Example sine synthesizer that creates a sine wave at the given input frequency.
+sine :: (VividAction m) => Cell (HandlingStateT m) Double ()
+sine = proc frequency -> do
```

Comment:
@miguel-negrao do you think this is an ok example? Or does it have any obvious mistakes?

Reply:
Yes, I don't think that works. You will be changing the synthdef all the time, resending it every time the Double value changes, which not only is not efficient but would cause the fade-in and fade-out process to happens all the time. Probably it would sound bad. You need to create a parameter named "frequency" in the synthdef.

Reply:
Also, no sound will play, as you the synthedef doesn't have an "out" ugen. I can correct this if you want.

Reply:
I think this would be the most minimalist example that still works well:

```
sine :: (VividAction m) => Cell (HandlingStateT m) Float ()
sine = proc frequency -> do
    liveSynth -< ((1 :: I "gate", 2::I "fadeSecs", realToFrac frequency:: I "freq"),
        out 0 [envGate ~* sinOsc (freq_ (V::V "freq"))], Started)
    returnA -< ()
```

Without envGate synths are not freed when the handlingstate calls release.

Tested with

```
liveProgram :: LiveProgram (HandlingStateT IO)
liveProgram = liveProgramMidiIO midiDevices midiDevices $ proc (_, _, _, midiMessage) -> do
    slider <- midiInHoldC 1.0 $ akaiMidiMix ^. #sl . ix' @0 -< midiMessage
    liftHandlingState sine' -< slider * 5000
```

Reply:
Weird, I get:

```
• Could not deduce (ToSig busNum0 '["gate", "fadeSecs", "freq"])
    arising from a use of ‘out’
  from the context: VividAction m
    bound by the type signature for:
               LiveCoding.Vivid.sine :: forall (m :: * -> *).
                                        VividAction m =>
                                        Cell (HandlingStateT m) Double ()
    at /home/turion/haskell/essence-of-live-coding/essence-of-live-coding-vivid/src/LiveCoding/Vivid.hs:97:1-60
  The type variable ‘busNum0’ is ambiguous
  These potential instances exist:
    instance ToSig Int16 args -- Defined in ‘Vivid.SynthDef.ToSig’
    instance ToSig Int32 args -- Defined in ‘Vivid.SynthDef.ToSig’
    instance ToSig Int64 args -- Defined in ‘Vivid.SynthDef.ToSig’
    ...plus 9 others
    ...plus 7 instances involving out-of-scope types
    (use -fprint-potential-instances to see them all)
• In the expression:
    out 0 [envGate ~* sinOsc (freq_ (V :: V "freq"))]
  In the expression:
    ((1 :: I "gate", 2 :: I "fadeSecs", 
      realToFrac frequency :: I "freq"), 
     out 0 [envGate ~* sinOsc (freq_ (V :: V "freq"))], Started)
  In the command: liveSynth -< ((1 :: I "gate", 2 :: I "fadeSecs", 
                                 realToFrac frequency :: I "freq"), 
                                out 0 [envGate ~* sinOsc (freq_ (V :: V "freq"))], Started)
```

Reply:
I could solve it by saying `out (0 :: Int) ...`

Reply:
I updated it, I think this should work then.

Reply:
That is strange indeed. Will have to check why the type annotation is needed in that case. Also, [I](https://hackage.haskell.org/package/vivid-0.5.1.0/docs/Vivid-SynthDef-TypesafeArgs.html#t:I) wraps a Float, so perhaps the Cell input should also be Float.

Reply:
Btw, the objective of my embedded approach is to remove the need of having to define a synth parameter just to pass a value from the cell input, like we do in this example, such that it looks almost like your first example attempt.

Reply:
I simplified the `Float` thing. For now I'm happy with the type annotation.

Reply:
> Btw, the objective of my embedded approach is to remove the need of having to define a synth parameter just to pass a value from the cell input, like we do in this example, such that it looks almost like your first example attempt.

I'll have to read up on that eventually :) If you think it's something we could merge into eolc, feel free to open a PR once this one is merged.